chain/watcher.go

```diff
@@ -51,22 +60,42 @@ type Watcher struct {
 func (c *Watcher) Run(ctx context.Context) error {
 	// init the done channel for each run since jobs may be started and stopped.
 	c.done = make(chan struct{})
+
+	// create a worker pool with workers to index tipsets as they become avaiable.
+	c.pool = workerpool.New(c.poolSize)
+
+	// create a tipset notifier, register it to observe tipset applications and set its current head.
+	notifier := &TipSetObserver{bufferSize: c.bufferSize}
+	head := c.api.Observe(notifier)
+	if err := notifier.SetCurrent(ctx, head); err != nil {
+		return err
+	}
+
+	// warm the tipset cache with confidence tipsets
+	if err := c.cache.Warm(ctx, head, c.api.ChainGetTipSet); err != nil {
+		return err
+	}
+
 	defer func() {
-		// ensure we shut down the pool when the watcher stops.
-		c.pool.Stop()
 		// ensure we clear the fatal error after shut down, this allows the watcher to be restarted without reinitializing its state.
 		c.setFatalError(nil)
+		// ensure we shut down the pool when the watcher stops.
+		c.pool.Stop()
+		// ensure we reset the tipset cache to avoid process stale state if watcher is restarted.
+		c.cache.Reset()
+		// unregister the observer
+		// TODO https://github.com/filecoin-project/lotus/pull/8441
+		//c.api.Unregister(notifier)
```

Comment:
This PR is merged now. Does this need to change?

Reply:
Not yet, it will change when we upgrade to the lotus release with the change, so 1.15.2 or something. (Currently, this exits in lotus master, but we don't use master as our dep)